src/test/regress/citus_tests/test/test_other_databases.py

```diff
@@ -0,0 +1,140 @@
+def test_main_commited_outer_not_yet(cluster):
+    c = cluster.coordinator
+
+    # create a non-main database
+    c.sql("CREATE DATABASE db1")
+
+    # we will use cur1 to simulate non-main database user and
+    # cur2 to manually do the steps we would do in the main database
+    with c.cur(dbname="db1") as cur1:
+        with c.cur() as cur2:
+
+            # let's start a transaction and find its transaction id
+            cur1.execute("BEGIN")
+            cur1.execute("SELECT txid_current()")
+            txid = cur1.fetchall()
+
+            # using the transaction id of the cur1 simulate the main database commands manually
+            cur2.execute("BEGIN")
+            cur2.execute(
+                "SELECT citus_internal.start_management_transaction('"
+                + str(txid[0][0])
+                + "')"
+            )
+            cur2.execute(
+                "SELECT citus_internal.execute_command_on_remote_nodes_as_user('CREATE USER u1;', 'postgres')"
+            )
+            cur2.execute(
+                "SELECT citus_internal.mark_object_distributed(1260, 'u1', 123123)"
+            )
+            cur2.execute("COMMIT")
+
+            # run the transaction recovery
+            c.sql("SELECT recover_prepared_transactions()")
+
+            # user should not be created on the worker because outer transaction is not committed yet
+            before_commit = c.sql_value(
+                """
+                SELECT result
+                FROM run_command_on_workers($$
+                    SELECT count(*) FROM pg_roles
+                    WHERE rolname = 'u1'
+                $$) LIMIT 1
+                """
+            )
```

Comment:
nit: if you're doing the limit one to only check a single worker anyway, I think it's easier to simply connect straight to the worker.

```suggestion
            before_commit = cluster.worker[0].sql_value(
                "SELECT count(*) FROM pg_roles WHERE rolname = 'u1'"
            )
```